src/nfts/vendors/puzzle.ts

```diff
@@ -0,0 +1,83 @@
+import { DataTransactionEntryString } from '@waves/ts-types/src/parts';
+
+import {
+  dataEntriesToRecord,
+  fetchDataEntries,
+} from '../../nodeApi/dataEntries';
+import {
+  CreateParams,
+  FetchInfoParams,
+  Nft,
+  NftAssetDetail,
+  NftVendor,
+  NftVendorId,
+} from '../types';
+
+interface PuzzleNftInfo {
+  id: string;
+  vendor: NftVendorId.Puzzle;
+  image: string;
+  creator: string;
+}
+
+const PUZZLE_MARKET_DAPP = '3PFQjjDMiZKQZdu5JqTHD7HwgSXyp9Rw9By';
+
+function puzzleDataUrl(nodeUrl: string) {
+  return new URL(`addresses/data/${PUZZLE_MARKET_DAPP}`, nodeUrl).toString();
+}
+
+function nftPropertyKey(id: string, property: string) {
+  return `nft_${id}_${property}`;
+}
+
+export class PuzzleNftVendor implements NftVendor<PuzzleNftInfo> {
+  id = NftVendorId.Puzzle as const;
+
+  is(nft: NftAssetDetail): boolean {
+    return nft.issuer === PUZZLE_MARKET_DAPP;
+  }
+
+  fetchInfo({ nfts, nodeUrl }: FetchInfoParams) {
+    if (nfts.length === 0) {
+      return [];
+    }
+
+    const nftIds = nfts.map(nft => nft.assetId);
+
+    return fetchDataEntries<DataTransactionEntryString>(
+      puzzleDataUrl(nodeUrl),
+      nftIds
+        .map(id => [nftPropertyKey(id, 'image'), nftPropertyKey(id, 'issuer')])
+        .flat()
```

Comment:
[`flatMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) can be used instead of `map` + `flat`